### PR TITLE
Load visual elements of contact-content pane incrementally

### PR DESCRIPTION
**Pull request by alxndrsn**
See #4982 for explanation of this work.

Issue: #5086

### CODE REVIEW

**Comment by alxndrsn** [comment]
I recommend reviewing changes at https://github.com/medic/medic-webapp/pull/5069/files?w=1

File: webapp/src/js/services/contact-view-model-generator.js [hunk 1]
```diff
@@ -272,18 +268,43 @@ angular.module('inboxServices').factory('ContactViewModelGenerator',
...
+              })
+              .then(function(reports) {
+                model.reports = reports;
+                model.loadingReports = false;
+              })
+              .catch(function(err) {
+                reject(err);
+                throw err;
+              });
+          });
```
Comment (garethbowen) [inline webapp/src/js/services/contact-view-model-generator.js]: I don't think you need to wrap this in a new promise... why not:

```js
model.reportLoader = loadChildren(model)
  .then(function(children) {
    model.children = children;
    model.loadingChildren = false;
    return loadReports(model);
  })
  .then(function(reports) {
    model.reports = reports;
    model.loadingReports = false;
    return reports;
  });
```

File: webapp/src/js/controllers/contacts.js [hunk 2]
```diff
@@ -209,52 +209,69 @@ var _ = require('underscore'),
...
-                  icon: xForm.icon,
-                  showUnmuteModal: showUnmuteModal(xForm.internalId)
+
+          $scope.setRightActionBar({
+            relevantForms: [], // REVIEWER: this disables the "New Action" button in action bar until full load is complete
+            selected: [selectedDoc],
+            sendTo: selectedDoc.type === 'person' ? selectedDoc : '',
+            canDelete: false, // REVIEWER: this disables the "Delete" button in action bar until full load is complete
+            canEdit: canEdit,
+          });
```
Comment (garethbowen) [inline webapp/src/js/controllers/contacts.js]: These REVIEWER comments make sense to me. I think it'd be worth leaving them there as code comments.

**Comment by garethbowen** [comment]
Also, the designers have requested a couple of mitigations for the content scrolling problem, documented on [the issue](https://github.com/medic/medic-webapp/issues/5086).

File: webapp/src/js/controllers/contacts.js [hunk 3]
```diff
@@ -209,52 +209,69 @@ var _ = require('underscore'),
...
+
+          return selected.reportLoader.then(function() {
+            return $q.all([
+              ContactSummary(selected.doc, selected.reports, selected.lineage),
+              Settings()
+            ])
+            .then(function(results) {
```
Comment (kennsippell) [inline webapp/src/js/controllers/contacts.js]: Consider improving readibility by replacing `results` here with `[summary, settings]` in the argument list for the function.

File: webapp/tests/karma/unit/services/contact-view-model-generator.js [hunk 4]
```diff
@@ -107,14 +109,24 @@ describe('ContactViewModelGenerator service', () => {
     KarmaUtils.restore(GetDataRecords);
   });
 
+   // REVIEWER: I suspect this should be implemented using Q, but I can't work out how
+  function waitForModelToLoad(model) {
```
Comment (kennsippell) [inline webapp/tests/karma/unit/services/contact-view-model-generator.js]: I've hit this problem twice now testing with karma. Unless there is actually a good better solution to this, maybe generalize it and put it somewhere shared? I'd use it.

`function waitForQPromise(promise)` ?

Comment (alxndrsn) [reply webapp/tests/karma/unit/services/contact-view-model-generator.js]: Have you done similar elsewhere?  If you have other examples, we can pull code out and share it.

Comment (kennsippell) [reply webapp/tests/karma/unit/services/contact-view-model-generator.js]: I was able to fix it each time by refactoring - either injecting `Promise` in place ot `Q` or using the `Q.deferred` interface. I never found a solution like yours.

**Comment by alxndrsn** [comment]
Replaced by #5095

**Pull request closed without merging**